jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsConnector.scala

```diff
@@ -56,26 +57,40 @@ private[jms] trait JmsConnector[S <: JmsSession] {
   override def preStart(): Unit = {
     // keep two elements since the time between initializing and connected can be very short.
     // always drops the old state, and keeps the most current (two) state(s) in the queue.
-    val pair = Source.queue[InternalConnectionState](2, OverflowStrategy.dropHead).preMaterialize()(this.materializer)
-    connectionStateQueue = pair._1
-    connectionStateSourcePromise.complete(Success(pair._2))
+    val (queue, source) =
+      Source
+        .queue[InternalConnectionState](2, OverflowStrategy.dropHead)
+        .toMat(BroadcastHub.sink(1))(Keep.both)
```

Comment:
Is it possible that a buffer of 1 could result in a strange sequence of connection states downstream?  The first subscriber should be alright, but for subsequent subscriptions it's possible they could get an old stale connection state from the `BroadcastHub.sink` buffer first before the latest messages from the queue.  

Also, if the queue overflows it's possible you could still see gaps in downstream consumers, but I guess that's how it worked before as well.

Reply:
Nice catch; When I designed this originally, I did this on purpose so that a state consumer would still get all the intermediate states if they subscribed "a bit late", i.e. after the connection has already been established. I also thought that the connection status stream wouldn't be one with high volume of messages, and consumers would probably easily keep up with the update rate.

The only easy way I saw feasible for draining the connection state was to use a BroadcastHub. This now means that in the case we have multiple status update consumers (which wasn't possible before), with one of those consumers being very, very slow (i.e. slower than the connection status change rate), there could be gaps in the status update sequence. 

We could increase the buffer sizes to mitigate this to a certain point, or even make the buffer size configurable. What do you think? Which option would you favor?

Reply:
I think keeping the buffer size of the broadcast hub to 1 mitigates the problem the best (there will only ever be possibly 1 stale element).  I wrote some code to demonstrate:

```
      val (queue, source) =
        Source
          .queue[Int](2, OverflowStrategy.dropHead)
          .toMat(BroadcastHub.sink(1))(Keep.both)
          .run()(this.materializer)
      val offers = (1 to 10).map(queue.offer)
      Await.result(Future.sequence(offers), scala.concurrent.duration.Duration.Inf)
      val subscriber1 = source.map(i => s"subscriber1: $i").runForeach(println)
// prints
subscriber1: 1
subscriber1: 9
subscriber1: 10
```

I can't think of a simple way to give the broadcast hub the same semantics as the queue (only buffer the last 2 elements) using the akka streams operator DSL.

Reply:
Yes, fully agree. And to continue your example, we could mitigate the gaps if the queue had a buffer size of 9 elements. So increasing the queue connection buffer size would mitigate that effect to a certain extent.

Reply:
Increasing the buffer will only serve to distribute more stale records for late subscribers.  For example, if we increased the buffer to 10, and 100 connection status events occur before a new subscriber begins consuming elements, then they will receive 10 of the earliest records, followed by whatever comes out of the queue `1,2,3,4,5,6,7,8,9,10,99,100`.

I don't know if it's a big deal or not to get stale events.  I assume it's not since the overflow strategy of the queue was `DropHead` and we could experience lossiness anyway.

Reply:
It is correct that there will always be a point where the stream will become lossy; if we buffer `n` elements in the queue with `DropHead` (that drops the oldest element from the buffer) and have a `BroadcastHub.sink(1)`, we will have one very old status update (stored in the broadcast hub), and the `n` most recent status updates. So with `10` elements and buffer size of `5`, we'll get `1,6,7,8,9,10` (with `10` being the most recent update). But still, event when we increase the buffer, eventually the stream will be losing updates. And if processing updates is slow, buffering more updates might not be the right mitigation. Therefore I was thinking that making it configurable could be an option.